File: test/.work/robust_size_limits/ts/sb/struct_blob_wrap.ts
import * as _ from "./_"
import * as Enum from "./enum"

export interface BlobWrap extends _.Serializable, _.Deserializable {
    textValue: string;
    binValue: Uint8Array;
    nums: number[];
    level: Enum.Level;
}

export const newBlobWrap = (): BlobWrap => {
    const s = {
        textValue: "",
        binValue: new Uint8Array(0),
        nums: [],
        level: 0,
    } as any as BlobWrap;
    s.set = (buf: _.Buffer) => setBlobWrap(buf, s);
    s.get = (buf: _.Buffer) => {
        const [res, err] = getBlobWrap(buf);
        if (err === null) Object.assign(s, res);
        return err;
    };
    return s;
}

export const eqBlobWrap = (a: BlobWrap, b: BlobWrap): boolean => {
    if (a === b) return true;
    if (a === null || b === null) return false;
    if (!_.eqText(a.textValue, b.textValue)) return false;
    if (!_.eqBin(a.binValue, b.binValue)) return false;
    if (!_.eqU8List(a.nums, b.nums)) return false;
    if (a.level !== b.level) return false;
    return true;
}

export const getBlobWrap = (buf: _.Buffer): [BlobWrap, Error | null] => {
    const s = newBlobWrap();
    const bitmaskSize = Math.ceil(4 / 8);
    const [bits, err] = buf.read(bitmaskSize);
    if (err !== null) return [s, err];
    if (_.GetBit(bits, 0)) {
        const [v, err] = _.getText(buf);
        if (err !== null) return [s, err];
        s.textValue = v;
    }
    if (_.GetBit(bits, 1)) {
        const [v, err] = _.getBin(buf);
        if (err !== null) return [s, err];
        s.binValue = v;
    }
    if (_.GetBit(bits, 2)) {
        const [v, err] = _.getU8List(buf);
        if (err !== null) return [s, err];
        s.nums = v;
    }
    if (_.GetBit(bits, 3)) {
        const [v, err] = _.getU8(buf);
        if (err !== null) return [s, err];
        s.level = v as any;
        if (!_.IsLevel(s.level as any)) return [s, new Error("get BlobWrap Level: invalid enum value")];
    }
    return [s, null];
}

export const setBlobWrap = (buf: _.Buffer, s: BlobWrap): Error | null => {
    if (s === null || s === undefined) return new Error(`set BlobWrap: value is null or undefined`);
    const bits = new Uint8Array(Math.ceil(4 / 8));
    const body = new _.Buffer();
    if (!_.eqText(s.textValue, "")) {
        const err = _.setText(body, s.textValue);
        if (err !== null) return err;
        _.SetBit(bits, 0, true);
    }
    if (!_.eqBin(s.binValue, new Uint8Array(0))) {
        const err = _.setBin(body, s.binValue);
        if (err !== null) return err;
        _.SetBit(bits, 1, true);
    }
    if (s.nums && s.nums.length > 0) {
        const err = _.setU8List(body, s.nums);
        if (err !== null) return err;
        _.SetBit(bits, 2, true);
    }
    if ((s.level as any) !== 0) {
        if (!_.IsLevel(s.level as any)) return new Error("set BlobWrap Level: invalid enum value");
        const err = _.setU8(body, s.level as any);
        if (err !== null) return err;
        _.SetBit(bits, 3, true);
    }

    const errBits = buf.write(bits);
    if (errBits !== null) return errBits;
    return buf.write(body.bytes);
}

export const getBlobWrapList = (buf: _.Buffer): [BlobWrap[], Error | null] => _.getList(buf, getBlobWrap);
export const setBlobWrapList = (buf: _.Buffer, v: BlobWrap[]): Error | null => _.setList(buf, v, setBlobWrap);
export const eqBlobWrapList = (a: BlobWrap[], b: BlobWrap[]): boolean => _.eqList(a, b, eqBlobWrap);

File: test/.work/robust_size_limits/go/sb/struct_blob_wrap.go
package sb

import (
	"bytes"
	"fmt"
	"math"
	"slices"
)

type BlobWrap struct {
	TextValue string  
	BinValue []byte  
	Nums []uint8  
	Level Level  
}

func GetBlobWrap(buf *bytes.Buffer, s *BlobWrap) error {
	if s == nil { return nil }
	bitSize := int(math.Ceil(float64(4) / 8.0))
	if buf.Len() < bitSize { return fmt.Errorf("GetBlobWrap bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetBlobWrap TextValue: %w", err) }
		s.TextValue = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetBin(buf)
		if err != nil { return fmt.Errorf("GetBlobWrap BinValue: %w", err) }
		s.BinValue = val
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetU8List(buf)
		if err != nil { return fmt.Errorf("GetBlobWrap Nums: %w", err) }
		s.Nums = val
	}
	if GetBit(bits, uint8(3)) {
		val, err := GetLevel(buf)
		if err != nil { return fmt.Errorf("GetBlobWrap Level: %w", err) }
		s.Level = val
		if !IsLevel(s.Level) { return fmt.Errorf("GetBlobWrap Level: 非法枚举值: %d", s.Level) }
	}
	if err := ValidateBlobWrap(s); err != nil { return fmt.Errorf("ValidateBlobWrap: %w", err) }
	return nil
}

func SetBlobWrap(buf *bytes.Buffer, s *BlobWrap) error {
	if s == nil { return fmt.Errorf("SetBlobWrap: nil value") }
	if err := ValidateBlobWrap(s); err != nil { return fmt.Errorf("ValidateBlobWrap: %w", err) }
	bits := make([]byte, uint8(math.Ceil(float64(4)/8.0)))
	body := bytes.NewBuffer(nil)
	if s.TextValue != "" {
		if err := SetText(body, s.TextValue); err != nil { return fmt.Errorf("SetBlobWrap TextValue: %w", err) }
		SetBit(bits, uint8(0), true)
	}
	if s.BinValue != nil {
		if err := SetBin(body, s.BinValue); err != nil { return fmt.Errorf("SetBlobWrap BinValue: %w", err) }
		SetBit(bits, uint8(1), true)
	}
	if len(s.Nums) > 0 {
		if err := SetU8List(body, s.Nums); err != nil { return fmt.Errorf("SetBlobWrap Nums: %w", err) }
		SetBit(bits, uint8(2), true)
	}
	if s.Level != 0 {
		if err := SetLevel(body, s.Level); err != nil { return fmt.Errorf("SetBlobWrap Level: %w", err) }
		SetBit(bits, uint8(3), true)
	}

	if _, err := buf.Write(bits); err != nil { return fmt.Errorf("SetBlobWrap write bitmask: %w", err) }
	_, err := body.WriteTo(buf); return err
}

func ValidateBlobWrap(s *BlobWrap) error {
	if s == nil { return nil }
	if s.Level != 0 && !IsLevel(s.Level) { return fmt.Errorf("Level 非法枚举值: %d", s.Level) }
	return nil
}

func EqBlobWrap(a, b *BlobWrap) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqText(a.TextValue, b.TextValue) { return false }
	if !EqBin(a.BinValue, b.BinValue) { return false }
	if !EqU8List(a.Nums, b.Nums) { return false }
	if a.Level != b.Level { return false }
	return true
}

// Standalone functions
func ReadBlobWrap(buf *bytes.Buffer) (*BlobWrap, error) {
	s := new(BlobWrap)
	return s, GetBlobWrap(buf, s)
}

type BlobWrapList []*BlobWrap
func GetBlobWrapList(buf *bytes.Buffer) (BlobWrapList, error) { return getList[*BlobWrap, BlobWrapList](buf, ReadBlobWrap) }
func SetBlobWrapList(buf *bytes.Buffer, v BlobWrapList) error { return setList(buf, v, SetBlobWrap) }
func ValidateBlobWrapList(v BlobWrapList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("BlobWrapList[%d]: nil item", i) }
		if err := ValidateBlobWrap(item); err != nil { return fmt.Errorf("BlobWrapList[%d]: %w", i, err) }
	}
	return nil
}
func EqBlobWrapList(a, b BlobWrapList) bool { return slices.EqualFunc(a, b, EqBlobWrap) }
